src/backend/cdb/motion/cdbmotion.c

```diff
@@ -1274,3 +1279,113 @@ UpdateSentRecordCache(int32 *sent_record_typmod)
 		*sent_record_typmod = NextRecordTypmod;
 	}
 }
+
+static bool
+IsICTypeExist(GpVars_Interconnect_Type type)
+{
+	for (int i = 0; i < CurrentIPCLayerImplNum; ++i)
+	{
+		if (IPCLayerImpls[i]->ic_type == type)
+			return true;
+	}
+
+	return false;
+}
+
+/*
+ * try best to find the most appropriate IPC layer implement.
+ */
+static int
+TryFindICType()
+{
+	/* IPCLayerImpls[0]->ic_type is valid at least */
+	Assert(CurrentIPCLayerImplNum);
+
+	int candidate_types_order[IPCLAYER_CANDIDATE_NUM] =
+		{Gp_interconnect_type, INTERCONNECT_TYPE_UDPIFC, IPCLayerImpls[0]->ic_type};
+
+	for (int i = 0; i < IPCLAYER_CANDIDATE_NUM; ++i)
+	{
+		if (IsICTypeExist(candidate_types_order[i]))
+			return candidate_types_order[i];
+	}
+
+	/* never run here, just keep compiler slience */
+	Assert(0);
+	return INTERCONNECT_TYPE_UDPIFC;
+}
+
+void
+SetCurrentMotionIPCLayer(int new_type)
+{
+	int type;
+
+	/*
+	 * Only set Gp_interconnect_type before interconnect.so loaded.
+	 */
+	if (!process_shared_preload_libraries_done)
+	{
+		Gp_interconnect_type = new_type;
+		return;
+	}
+
+	if (IsICTypeExist(new_type))
+	{
+		type = new_type;
+	}
+	else
+	{
+		if (!IsUnderPostmaster)
+		{
+			/* postmaster process */
+			ereport(WARNING,
+					(errcode(ERRCODE_WARNING_GP_INTERCONNECTION),
+					 errmsg("No IPC layer implement found with type: %d, "
+							"choose one from the loaded implements.",
+							new_type)));
+
+			type = TryFindICType();
+		}
+		else
+		{
+			/* backend process */
+			ereport(ERROR,
+					(errcode(ERRCODE_GP_INTERCONNECTION_ERROR),
+					 errmsg("No IPC layer implement found with type: %d.",
+							new_type)));
+		}
+	}
+
+	for (int i = 0; i < CurrentIPCLayerImplNum; ++i)
+	{
+		if (IPCLayerImpls[i]->ic_type == type)
+		{
+			CurrentMotionIPCLayer = IPCLayerImpls[i];
+			Gp_interconnect_type = CurrentMotionIPCLayer->ic_type;
+			break;
+		}
+	}
+
+	return;
+}
+
+void
+RegisterIPCLayerImpl(MotionIPCLayer *impl)
+{
+	if (CurrentIPCLayerImplNum >= INTERCONNECT_TYPE_NUM)
+	{
+		ereport(WARNING,
+				(errcode(ERRCODE_WARNING_GP_INTERCONNECTION),
+				 errmsg("There is no space for a new IPC layer implement.")));
+
+		return;
+	}
+
+	if (IsICTypeExist(impl->ic_type))
+	{
+		elog(LOG, "ic_type: %d has been registered.", impl->ic_type);
```

Comment:
ERROR

Reply:
fix it with [commit](https://github.com/cloudberrydb/cloudberrydb/pull/646/commits/cc3a12c740074a515a04896ea27a63cdf46ace57)